dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/resolver/CategoryOptionGroupResolver.java

```diff
@@ -28,154 +28,95 @@
-     * COCUID3 and COCUID4 are resolved from COGUID2
-     *
-     * DEUID = Data Element UID COCUID = Category Option Combo UID COGUID =
-     * Category Option Group UID
-     *
-     * @param expression a Data Element Expression
-     * @return an expression containing additional CategoryOptionCombos based on
-     *         the given Category Option Group
-     */
     @Override
-    @Transactional( readOnly = true )
     public String resolve( String expression )
     {
-        // Get a DimensionalItemId from the expression. The expression is parsed
-        // and
-        // each element placed in the DimensionalItemId
         Set<DimensionalItemId> dimItemIds = expressionService.getExpressionDimensionalItemIds( expression,
             INDICATOR_EXPRESSION );
-        List<String> resolvedOperands = new ArrayList<>();
-        if ( isDataElementOperand( dimItemIds ) )
-        {
-            DimensionalItemId dimensionalItemId = dimItemIds.stream().findFirst().get();
-            // First element is always the Data Element Id
-            String dataElementUid = dimensionalItemId.getId0();
-
-            resolvedOperands
-                .addAll( evaluate( dataElementUid, dimensionalItemId.getId1(), dimensionalItemId.getId2() ) );
 
-            resolvedOperands.addAll( evaluate( dataElementUid, dimensionalItemId.getId2(), null ) );
-        }
-        if ( resolvedOperands.isEmpty() )
+        for ( DimensionalItemId id : dimItemIds )
         {
-            // nothing to resolve, add the expression as it is
-            resolvedOperands.add( expression );
+            if ( id.getItem() != null && id.getId1() != null && id.getId1().contains( CATEGORY_OPTION_GROUP_PREFIX ) )
+            {
+                String[] cogUidList = id.getId1().split( LOGICAL_AND );
+                expression = getExpression( expression, id, cogUidList );
+            }
         }
-        return Joiner.on( "+" ).join( resolvedOperands );
-    }
 
-    private List<String> evaluate( String dataElementUid, String uid, String uid2 )
-    {
-        List<String> resolvedExpression = new ArrayList<>();
-        Optional<String> cogUid = getCategoryOptionGroupUid( uid );
-        if ( cogUid.isPresent() )
-        {
-            Set<String> cocs = resolveCoCFromCog( cogUid.get() );
-            resolvedExpression = Arrays.asList( resolve( cocs, dataElementUid, uid2 ).split( "\\+" ) );
-        }
-        return resolvedExpression;
+        return expression;
     }
 
-    private String resolve( Set<String> cocs, String dataElementUid, String third )
+    private String getExpression( String expression, DimensionalItemId id, String[] cogUidList )
     {
-        boolean isAoc = isAoc( third );
-
-        return cocs.stream()
-            .map( coc -> EXP_OPEN + dataElementUid + SEPARATOR + coc + (isAoc ? SEPARATOR + third : "") + EXP_CLOSE )
-            .collect( Collectors.joining( "+" ) );
-    }
+        for ( String cogUid : cogUidList )
```

Comment:
When a coGroup: tag is found, you need to
1. Find the set of CategoryOptionCombos for each CategoryOptionGroup
2. Find the intersection of these CategoryOptionCombo sets.
3. Use this intersection to expand the DimensionalItemId.

The code currently does the effective union (without removing duplicates), not the intersection.

If there are no CategoryOptionCombos in the intersection, then I suggest not doing the replacement. The DimensionalItemId will not be looked up for metadata, and no value will be found for the expression. I think it will do the right thing.

Reply:
done